subsys/mpsl/init/mpsl_init.c

```diff
@@ -345,6 +349,80 @@ static void mpsl_calibration_work_handler(struct k_work *work)
 }
 #endif /* CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC && !CONFIG_SOC_SERIES_NRF54HX */
 
+#if !defined(CONFIG_SOC_SERIES_NRF54HX) && !defined(CONFIG_SOC_SERIES_NRF54LX)
+static uint16_t m_measure_hfxo_time_us(void)
+{
+	bool clock_running_on_boot = false;
+
+	if (nrf_clock_hf_is_running(NRF_CLOCK, NRF_CLOCK_HFCLK_HIGH_ACCURACY)
+		|| nrf_clock_start_task_check(NRF_CLOCK, NRF_CLOCK_DOMAIN_HFCLK)) {
+		clock_running_on_boot = true;
+	}
+
+	/* Make sure the HFCLK is off. */
+	nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_HFCLKSTOP);
+
+	nrf_timer_prescaler_set(NRF_TIMER0,
+		NRF_TIMER_PRESCALER_CALCULATE(NRF_TIMER_BASE_FREQUENCY_GET(NRF_TIMER0),
+			1000000));
+
+	nrf_timer_cc_set(NRF_TIMER0, 0, 400);
+	nrf_timer_event_clear(NRF_TIMER0, NRF_TIMER_EVENT_COMPARE0);
+	nrf_timer_task_trigger(NRF_TIMER0, NRF_TIMER_TASK_CLEAR);
+	nrf_timer_task_trigger(NRF_TIMER0, NRF_TIMER_TASK_START);
+
+	while (!nrf_timer_event_check(NRF_TIMER0, NRF_TIMER_EVENT_COMPARE0)) {
+#if defined(CONFIG_SOC_SERIES_BSIM_NRFXX)
+		k_busy_wait(10);
+#endif
+	}
+
+	nrf_clock_event_clear(NRF_CLOCK, NRF_CLOCK_EVENT_HFCLKSTARTED);
+
+	/* Measure how long it takes to get EVENT_HFCLKSTARTED.
+	 * This is larger than what the crystal needs, as it may also include
+	 * debounce periods implemented in hardware, depending on the product.
+	 */
+	nrf_timer_task_trigger(NRF_TIMER0, NRF_TIMER_TASK_CAPTURE0);
+
+	const uint32_t measurement_start = nrf_timer_cc_get(NRF_TIMER0, 0);
+
+	nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_HFCLKSTART);
+
+	while (!nrf_clock_event_check(NRF_CLOCK, NRF_CLOCK_EVENT_HFCLKSTARTED)) {
+#if defined(CONFIG_SOC_SERIES_BSIM_NRFXX)
+		k_busy_wait(10);
+#endif
+	}
+
+	nrf_timer_task_trigger(NRF_TIMER0, NRF_TIMER_TASK_CAPTURE0);
+
+	const uint32_t measurement_end = nrf_timer_cc_get(NRF_TIMER0, 0);
```

Comment:
The measured time was using internal HF RC, should we consider the accuracy of this clock over the temperature range into the measured value?

Reply:
It already is. Assuming 5000 ppm for HFINT

1400*(5000/1000000) = 7us (much less than the padding)